File: backend/src/main/scala/domains/blog/http/api/ListBlogs.scala
package domains.blog.http.api

import cats.effect.IO
import domains.auth.http.AuthenticatedApi
import domains.auth.objects.AuthUser
import domains.blog.http.BlogApiSupport.ListBlogsInput
import domains.blog.http.codec.BlogHttpCodecs.given
import domains.blog.objects.response.BlogListResponse
import domains.blog.table.blog.BlogPostQueryTable
import domains.user.objects.Username
import io.circe.Encoder
import org.http4s.{Method, Request, Status}
import shared.http.utils.PageRequestQuerySupport
import shared.http.{ApiPath, PathParams}

import java.sql.Connection

object ListBlogs extends AuthenticatedApi[ListBlogsInput, BlogListResponse]:

  override val method: Method = Method.GET
  override val path: ApiPath = ApiPath("/api/blogs")
  override val successStatus: Status = Status.Ok
  override protected val outputEncoder: Encoder[BlogListResponse] = summon[Encoder[BlogListResponse]]

  override def decode(request: Request[IO], pathParams: PathParams): IO[ListBlogsInput] =
    val _ = pathParams
    IO.pure(
      ListBlogsInput(
        authorUsername = request.uri.query.params.get("username").map(Username.canonical),
        pageRequest = PageRequestQuerySupport.parsePageRequest(request.uri.query.params)
      )
    )

  override def plan(connection: Connection, actor: AuthUser, input: ListBlogsInput): IO[BlogListResponse] =
    input.authorUsername match
      case Some(username) =>
        BlogPostQueryTable.listByAuthor(connection, username, actor.username, input.pageRequest.normalized)
      case None =>
        BlogPostQueryTable.listAll(connection, actor.username, input.pageRequest.normalized)

File: frontend/src/apis/blog/ListBlogs.ts
import type { APIMessage } from '@/system/api/api-message'
import type { BlogListResponse } from '@/objects/blog/response/BlogListResponse'
import type { PageRequest } from '@/objects/shared/PageRequest'
import type { Username } from '@/objects/user/Username'
import { usernameValue } from '@/objects/user/Username'

/** 查询博客列表；可按作者过滤并分页，输出博客摘要分页响应。 */
export class ListBlogs implements APIMessage<BlogListResponse> {
  declare readonly responseType?: BlogListResponse
  readonly method = 'GET'
  readonly apiPath: string

  constructor(authorUsername?: Username | null, pageRequest?: PageRequest) {
    const params = new URLSearchParams()
    if (authorUsername) {
      params.set('username', usernameValue(authorUsername))
    }
    if (pageRequest) {
      params.set('page', String(pageRequest.page))
      params.set('pageSize', String(pageRequest.pageSize))
    }
    this.apiPath = `blogs${params.size > 0 ? `?${params.toString()}` : ''}`
  }

  body(): undefined {
    return undefined
  }
}
